Adding `__repr__` for container workload classes

khoaguin opened the pull request:
Solving these issues:
- [x] [Add __repr__ for SyftImage related classes#905](https://github.com/OpenMined/Heartbeat/issues/905)
- [x] [Add __repr__ to SyftWorker and SyftWorkerPool#906](https://github.com/OpenMined/Heartbeat/issues/906)

Summary of what this PR does:
-  `<client>.images` should show a table containing only the main ones: `id, image_tag, image_hash, created_at`
- Doing `client.worker_pools` would be preferred over client.worker_pool.get_all()
- If `tag = "openmined/test-nginx:0.7.8"`, we can index worker image with `client.images[tag]`
-  `client.worker_pools` show a list of worker pools with `id, pool_name, workers, healthy (healthy/all), running (running/all), image tag/name, created_at`
- we can also index a worker pool with `client.worker_pools[tag]`
- visualising a worker pool, should show first the inline fields like in `client.worker_pools` and after a table of workers show workers' information
- `client.worker_pools[0].workers` shows all the worker `id, image tag/name, resources pool name, healthcheck (healthy/unhealthy), status (created/running/stopped), job`
- `worker = client.worker_pools[0].workers[0]`, then worker will show `id/name, image tag, worker pool name, healthy, status`

khoaguin committed:
[syft/worker_image] add `__repr__` and `__repr_attrs__` for `SyftWorkerImage`

khoaguin committed:
Merge branch 'container-workload-1' into feat/container-workload-classes-repr

khoaguin committed:
[syft/worker_image] Define repr and __repr_attrs__ for
`SyftWorkerImage`, `SyftWorkerImageTag`, `ContainerImageRegistry`

Co-authored-by: Peter Chung <peterchung241@gmail.com>

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

khoaguin commented:
Before building image
![image](https://github.com/OpenMined/PySyft/assets/88959106/b04c3c7d-c430-4f14-a0a8-9e7032434a37)

SyftWorkerImage repr
![image](https://github.com/OpenMined/PySyft/assets/88959106/3e433c44-ea5a-4e57-b6dd-b328ac6bd0ce)

After building image
![image](https://github.com/OpenMined/PySyft/assets/88959106/24362446-11e1-446b-8369-6e1fd1ced2d1)

khoaguin commented:
should we change something about this repr?
![image](https://github.com/OpenMined/PySyft/assets/88959106/a7cd5477-2f23-4930-8d73-4d31ef1bf108)

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/533cfc5b-4b62-4f12-933f-7ba20ef0ee0a)

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/ec1374ae-fe4e-404e-aa0a-ad9ba3162752)

khoaguin committed:
[syft/worker_image] add repr for `SyftWorker` and `WorkerPool`

Co-authored-by: Peter Chung <peterchung241@gmail.com>

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/a7509c8e-b0a1-4723-911e-2d158231bed0)

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/9dbfc235-7a77-4d55-be5f-a689bd3fb9de)

shubham3121 commented:
@khoaguin can you resolve merge conflict and merge with dev

shubham3121 commented on packages/syft/src/syft/protocol/protocol_version.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "1": {
+  "dev": {
```
I think the procotol_version.json seems incorrect, maybe to incorrect merge with dev. Can you reset this file with `dev` 
`git checkout dev -- packages/syft/src/syft/protocol/protocol_version.json` and then in a new terminal
`tox -e syft.protocol.check` to stage any protocol changes.

khoaguin replied:
I noticed there are some differences in my branch's `protocol_version.json` and `dev`'s `protocol_version.json`. Will simply copying the content of `dev`'s `protocol_version.json` into mine work?

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

khoaguin committed:
[syft/container_workload] add some prints in the container-images notebook
to show their `repr`

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

khoaguin committed:
[syft/worker_image] remove some unnecessary `__repr_html__`
add `images` and `worker_pools` properties for `DomainClient`
remove the `dockerfile_name` column in `SyftWorkerImage List` table

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/1826b248-e33f-40e0-883a-52c43a7d6dda)

shubham3121 committed:
fix protocol version json

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

khoaguin committed:
[syft/worker_image] - `WorkerImageService.get_all()` now returns a `DictTuple`
- Can get a list of images and worker pools with `client.images` or `client.worker_pools`
- Remove some unnecessary `__repr_html__` functions

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/1bf60bf9-fccb-4323-b587-bf7805518dfb)

khoaguin committed:
[syft/worker_image] done table of worker pool list

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/cccc5677-3ebc-4e41-8271-5c63bf6b66de)

khoaguin committed:
[syft/worker_image] index a worker pool with `client.worker_pools[tag]`

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/0436c783-bb01-4886-b182-855b31f8b255)
![image](https://github.com/OpenMined/PySyft/assets/88959106/6af598dd-b6b9-4639-a5b5-91a6ced68692)
![image](https://github.com/OpenMined/PySyft/assets/88959106/3015d13a-c136-46eb-a219-0f25bac494a5)
![image](https://github.com/OpenMined/PySyft/assets/88959106/e278edfb-cc34-4e29-aa22-b3ba26b80391)

khoaguin committed:
[syft/worker_image] show `WorkerPool` similarly to `Dataset`
- show a table of `SyftWorkers`

khoaguin committed:
[syft/worker_image] working on `client.workers`
- add some lines in notebook api to show new `repr`s

khoaguin committed:
[syft/worker_image] add TODO for `domain_client.workers`

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

khoaguin commented:
Changes after mergingin with `dev`: we have a new column for `Job` in `WorkerPool`'s repr

![image](https://github.com/OpenMined/PySyft/assets/88959106/6f405f5d-6d07-47d0-9b4e-c843018cca22)

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/c416c14e-2697-43b0-8e2c-d57e5f2191f0)

khoaguin committed:
update `protocol_version.json`

kiendang commented on packages/syft/src/syft/service/worker/worker_image.py:
```diff
@@ -91,6 +101,10 @@ class SyftWorkerImage(SyftObject):
     created_at: DateTime = DateTime.now()
     created_by: SyftVerifyKey
+    source_file: Optional[str]
+    full_tag_str: Optional[str]
```
nitpick: This could just be `full_tag`? The type is already `str`.

khoaguin replied:
Done

kiendang commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -43,6 +64,7 @@ class SyftWorker(SyftObject):
     image_hash: Optional[str]
     healthcheck: Optional[WorkerHealth]
     status: WorkerStatus
+    full_image_tag_str: Optional[str]
```
`full_image_tag`

khoaguin replied:
Done

kiendang commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -79,11 +107,76 @@ class WorkerPool(SyftObject):
+
+    @property
+    def running_workers(self) -> List[SyftWorker]:
+        return [
+            worker
+            for worker in self.workers
+            if get_worker_container_status(worker) == WorkerStatus.RUNNING
```
This would create multiple `DockerClient`s, one for each worker, but don't worry about this I'll push a fix for this.

kiendang committed:
Merge branch 'dev' into feat/container-workload-classes-repr

kiendang commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -99,3 +194,57 @@ class ContainerSpawnStatus(SyftBaseModel):
+)
+
+
+def _get_worker_container_status(
+    client: docker.DockerClient,
+    worker: SyftWorker,
+    container: Optional[Container] = None,
+) -> Union[Container, SyftError]:
+    if container is None:
+        container = _get_worker_container(client, worker)
+
+    if isinstance(container, SyftError):
+        return container
+
+    container_status = container.status
+
+    return _CONTAINER_STATUS_TO_WORKER_STATUS.get(
+        container_status,
+        SyftError(message=f"Unknown container status: {container_status}"),
+    )
```
@rasswanth-s

rasswanth-s replied:
Awesome 🙌

kiendang commented:
@khoaguin Just resolved conflicts with dev. Will take a look at the `DictTuple` thing.

kiendang commented on packages/syft/src/syft/client/api.py:
```diff
@@ -384,6 +384,9 @@ def _repr_html_(self) -> Any:
         results = self.get_all()
         return results._repr_html_()
 
+    def __len__(self) -> int:
+        return len(self.get_all())
+
```
I think we should not add this here. This is an expensive call to get `__len__` and it's not obvious to users. They could just do `len(client...get_all())` instead.

khoaguin replied:
I think this depends on the `API.get_all()`, but we already use it `_repr_html` and `__getitem__` so it gets called all the time
![image](https://github.com/OpenMined/PySyft/assets/88959106/244d0c7a-3762-488c-9c16-85cd1e377fb1)

khoaguin replied:
And I add this as it makes a lot of sense to do something like `len(client.images)` or `len(client.worker_pools)` to see how many are there

kiendang replied:
> I think this depends on the `API.get_all()`, but we already use it `_repr_html` and `__getitem__` so it gets called all the time

It's true. These are also expensive, and I plan to discuss with the product team about this. Doesn't mean we should just keep adding to it in the meantime though. Plus this is beyond the scope of this PR. You're not just adding to `client.images` or `client.worker_pools` but all other endpoints (e.g. `client.datasets`) as well.

khoaguin replied:
Thanks @kiendang. This makes sense. I removed it

khoaguin committed:
[syft/worker_image] add some statements to the
api notebook to show the new `repr`s

kiendang commented on packages/syft/src/syft/client/domain_client.py:
```diff
@@ -197,6 +198,35 @@ def code_history(self) -> CodeHistoriesDict:
+            return self.api.services.worker_pool
+        return None
+
+    @property
+    def workers(self):
+        """
+        TODO: We will work on this later with a new, separate WorkerStash
+        """
+        result = self.api.services.worker_pool.get_all()
+        if isinstance(result, DictTuple):
+            worker_names = ""
+            image_tag_workers = {}
+            for worker_pool in result:
+                for worker in worker_pool.workers:
+                    worker_names += f"{worker.name}\n"
+                image_tag_workers[worker_pool.syft_worker_image_name_tag] = worker_names
+            return DictTuple(image_tag_workers)
+        else:
+            return SyftError(message=f"{result.err()}")
```
It makes sense that `client.workers()` does not print out the `_repr_html_`. Plain `DictTuple` objects do not have the `_repr_html_` method.

Also if we're not using this right now I think it's better to remove it from this PR. We can add it back when we actually use it.

khoaguin replied:
Done

khoaguin committed:
[syft/worker_image] - remove unnecessary and ineffective code
- update `10-container-images` api notebook to reflect changes

khoaguin committed:
[syft/worker_image] refactoring some variable names

Co-authored-by: Kien Dang <mail@kien.ai>

khoaguin committed:
syncing with 0.8.3 packages/syft/src/syft/protocol/protocol_version.json

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

rasswanth-s commented on packages/syft/src/syft/service/worker/worker_image.py:
```diff
@@ -91,6 +101,10 @@ class SyftWorkerImage(SyftObject):
     created_at: DateTime = DateTime.now()
     created_by: SyftVerifyKey
+    source_file: Optional[str]
+    full_tag: Optional[str]
```
NIT: since already have a SyftWorkerImageTag, we could avoid creating full_tag attribute?

rasswanth-s replied:
to retrieve the full tag, we could do
image_tag.full_tag

khoaguin replied:
Done. Removed `full_tag` property and change `SyftWorkerImageTag` to `SyftWorkerImageIdentifier`

rasswanth-s commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -43,6 +65,7 @@ class SyftWorker(SyftObject):
     image_hash: Optional[str]
     healthcheck: Optional[WorkerHealth]
     status: WorkerStatus
+    full_image_tag: Optional[str]
```
MAYBE: instead of full_image_tag we could have it as image_tag?

rasswanth-s replied:
MAYBE: For image_tag, we could pass SyftWorkerImageTag class from SyftWorkerImage

khoaguin replied:
Done 
![image](https://github.com/OpenMined/PySyft/assets/88959106/c5964610-5018-42e0-8299-f1cd878f48a2)

kiendang commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -79,11 +109,76 @@ class WorkerPool(SyftObject):
+    ]
 
     name: str
     syft_worker_image_id: UID
+    syft_worker_image_name_tag: Optional[str]
```
This should just be `syft_worker_image_tag`. I also think just `worker_image_id` and `worker_image_tag` is better, or even just `image_id` and `image_tag`. @rasswanth-s what do you think?

rasswanth-s replied:
@kiendang  image_id and image_tag is more better 👍

khoaguin replied:
Now pass the whole `SyftWorkerImage` 
![image](https://github.com/OpenMined/PySyft/assets/88959106/2287be7a-e026-4993-bee3-f03eac8595d4)

rasswanth-s commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -79,11 +109,76 @@ class WorkerPool(SyftObject):
 
     name: str
     syft_worker_image_id: UID
+    syft_worker_image_name_tag: Optional[str]
```
MAYBE: we could name it `syft_worker_image_tag`  and we could initiliaze with SyftWorkerImageTag class to mantain consistency.

khoaguin replied:
Done. Now the `SyftWorkerImage` is passed
![image](https://github.com/OpenMined/PySyft/assets/88959106/492ad7f3-d1c9-4ac7-9ef2-679edf03c8ae)

rasswanth-s commented on packages/syft/src/syft/service/worker/worker_image_service.py:
```diff
@@ -108,16 +111,15 @@ def build(
-
-        return result.ok()
+        images: List[SyftWorkerImage] = result.ok()
+        return DictTuple((im.full_tag, im) for im in images)
```
Q: What is the functionality of DictTuple, even though , we have specific columns, I think __repr_attrs__ overrides as the full_tag field is not displayed.


<img width="963" alt="Screenshot 2023-12-22 at 10 44 56 AM" src="https://github.com/OpenMined/PySyft/assets/43314053/f6f0ca94-d2a7-43c6-9ac0-e98c64ff9a22">

rasswanth-s commented on packages/syft/src/syft/service/worker/worker_pool_service.py:
```diff
@@ -104,6 +108,7 @@ def create_pool(
             syft_worker_image_id=image_uid,
             max_count=number,
             workers=workers,
+            syft_worker_image_name_tag=worker_image.full_tag,
```
MAYBE: We could set the image tag to `worker_image.image_tag` ?

khoaguin replied:
Done. Now we pass `worker_image: SyftWorkerImage` into `WorkerPool`

rasswanth-s commented on packages/syft/src/syft/service/worker/worker_image.py:
```diff
@@ -91,6 +101,10 @@ class SyftWorkerImage(SyftObject):
     image_hash: Optional[str]
     created_at: DateTime = DateTime.now()
     created_by: SyftVerifyKey
+    source_file: Optional[str]
```
Q: Is source file attribute being used, I could find it being used?

khoaguin replied:
It's being used when user pass in a Dockerfile and now it holds the name of that file

kiendang replied:
Hmm I still don't see where this attribute is used?

khoaguin replied:
It's only used if we pass the a `.dockerfile`. For now in the api notebook we pass directly the string so we are not using it yet

kiendang replied:
If we're not using it yet we should not include it in this PR. We should only add it in when we use it and implement support for it.

shubham3121 replied:
This shouldn't be tracked at SyftWorkerImage level, this is more of a WorkerConfig thing.

khoaguin commented:
We have a new convention for syft image tag, repo, name and registry (after reading the [Docker docs](https://docs.docker.com/engine/reference/commandline/tag/#tag-an-image-referenced-by-name-and-tag) and disscuss with Rasswanth and Kien) like below: 
For example if we have the image identifier as `docker.io/openmined/test-nginx:0.7.8` then
```
registry = docker.io
repo = openmined/test-nginx
tag = 0.7.8
repo_with_tag = openmined/test-nginx:0.7.8
full_name = docker.io/openmined/test-nginx
full_name_with_tag = docker.io/openmined/test-nginx:0.7.8
```
This will reflect the following changes in the codebase:
- refactor `SyftWorkerImageTag` into `SyftWorkerImageIdentifier` since tag only refers to the version (e.g. `0.7.8`) and cause confusions
- replace `SyftWorkerImage.full_tag` with `SyftWorkerImageIdentifier.repo_with_tag` everywhere
- add the following properties to `SyftWorkerImageIdentifier`: `repo_with_tag`, `full_name_with_tag`

khoaguin committed:
[syft/worker_image] refactor `SyftWorkerImageTag` to `SyftWorkerImageIdentifier`

khoaguin commented:
@rasswanth-s @shubham3121 I need some inputs on this:  
Irina wants to have an empty list returned when the image is not built, but for now the way we build is to get the `image_id` first with 
```python
dockerfile_list = domain_client.images
workerimage = dockerfile_list[0]
domain_client.api.services.worker_image.build(uid=workerimage.id, tag=docker_tag)
```
So I patched it currently with return a list of images with empty `image_hash` and `image_identifier`
![image](https://github.com/OpenMined/PySyft/assets/88959106/4ff789b2-c669-4a37-b899-98625ae28169)

khoaguin commented:
Updated `client.images`
![image](https://github.com/OpenMined/PySyft/assets/88959106/b89a2216-3fb5-4d9d-b53c-ff284f17ad35)

khoaguin committed:
[syft/worker_image] change `SyftWorkerImageIdentifier.full_tag` to `repo_with_tag`

shubham3121 commented:
> @rasswanth-s @shubham3121 I need some inputs on this: Irina wants to have an empty list returned when the image is not built, but for now the way we build is to get the `image_id` first with
> 
> ```python
> dockerfile_list = domain_client.images
> workerimage = dockerfile_list[0]
> domain_client.api.services.worker_image.build(uid=workerimage.id, tag=docker_tag)
> ```
> 
> So I patched it currently with return a list of images with empty `image_hash` and `image_identifier` ![image](https://private-user-images.githubusercontent.com/88959106/292465324-4ff789b2-c669-4a37-b899-98625ae28169.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMyNDU4NDYsIm5iZiI6MTcwMzI0NTU0NiwicGF0aCI6Ii84ODk1OTEwNi8yOTI0NjUzMjQtNGZmNzg5YjItYzY2OS00YTM3LWI4OTktOTg2MjVhZTI4MTY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIyVDExNDU0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE2MmY2ZWQwZmVjZTc2YTQ2ODM0NDFjZGI1ZDlhODJjM2MxYmQ5NmRmMDU0NGQwOTBkOTUyYzM0OWQzZTY3MDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jk_aclJzQ4vKmZhhBFyksx7WRzLLBFkkyZJOv8QiVWA)

Yes that sounds good, in case the image_tag or image_hash is not available we should show them as None.

shubham3121 commented on packages/syft/src/syft/service/worker/utils.py:
```diff
@@ -62,25 +70,26 @@ def run_container_using_docker(
     if not orchestration == WorkerOrchestrationType.DOCKER:
         return SyftError(message="Only Orchestration via Docker is supported.")
 
-    with contextlib.closing(docker.from_env()) as client:
```
This seems inconsistent with `dev` branch. This line shouldn't be deleted, and the next loop should be within this scope.

khoaguin replied:
Thanks @shubham3121. It was added back

khoaguin committed:
[syft/worker_image] passing `SyftWorkerImage` into `WorkerPool` and `SyftWorker`

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

khoaguin commented:
Updated `client.worker_pools`
![image](https://github.com/OpenMined/PySyft/assets/88959106/324460ec-4cf3-4d21-b8f0-df3b11709f1b)

khoaguin commented:
Updated single worker pool repr
![image](https://github.com/OpenMined/PySyft/assets/88959106/1fe00b21-70ee-43f9-b9f1-470e7f7344b4)
Indexing with string using image repo and tag
![image](https://github.com/OpenMined/PySyft/assets/88959106/de989a8a-ee21-4613-8e0c-c2cb630edc3c)

khoaguin commented:
Showing a list of workers
![image](https://github.com/OpenMined/PySyft/assets/88959106/0cd8be0a-1d8c-4043-befa-2ee4be6d354a)

kiendang commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -78,12 +109,77 @@ class WorkerPool(SyftObject):
 
     name: str
-    syft_worker_image_id: UID
+    image_id: UID
+    image: SyftWorkerImage
```
`image: SyftWorkerImage` already contains `image_id` here (`image_id` here is the same as `image.id` right?) so `image_id: UID` is redundant here.

khoaguin replied:
That's true, thanks @kiendang. Updated the code

kiendang commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -43,6 +66,7 @@ class SyftWorker(SyftObject):
     image_hash: Optional[str]
     healthcheck: Optional[WorkerHealth]
     status: WorkerStatus
+    image: SyftWorkerImage
```
Similarly, we no longer need `image_hash: Optional[str]` here since `image: SyftWorkerImage` already contains it.

khoaguin replied:
That's true, thanks @kiendang . Updated the code

khoaguin committed:
Merge branch 'dev' into feat/container-workload-classes-repr

khoaguin committed:
[syft/worker_image] remove some unnecessary attributes
for `WorkerPool` and `SyftWorker`

Co-authored-by: Kien Dang <mail@kien.ai>

khoaguin committed:
[syft/worker_image] fix `test_nested_jobs` failed due to
updates in `SyftWorker`

khoaguin committed:
[syft/worker_image] update relevant code after removing
unnecessary attributes for `WorkerPool` and `SyftWorker`

khoaguin committed:
[syft/worker_image] making `image: SyftWorkerImage` an optional
attribute for `WorkerPool` and `SyftWorker`
- handle their `_coll_repr_` functions when `image` not exists

khoaguin committed:
[syft/worker_image] remove `image_hash` from SyftWorker's __attr_searchable__

shubham3121 committed:
Merge branch 'dev' into feat/container-workload-classes-repr

shubham3121 committed:
remove source_file field from SyftWorkerImage

shubham3121 commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -62,14 +85,30 @@ def get_job_repr(self):
-            "job": self.get_job_repr(),
-            "created at": str(self.created_at),
-        }
+    def get_status_healthcheck(self) -> None:
```
Health checks should be done via API call to the server not on the client side.

khoaguin replied:
Thanks @shubham3121, done!

shubham3121 commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -78,12 +117,85 @@ class WorkerPool(SyftObject):
+
+    @property
+    def running_workers(self) -> List[SyftWorker]:
+        with contextlib.closing(docker.from_env()) as client:
```
Same here, these should be triggered over an API call instead.

khoaguin replied:
Thanks @shubham3121, done!

shubham3121 commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -78,12 +117,85 @@ class WorkerPool(SyftObject):
+
+    @property
+    def healthy_workers(self) -> List[SyftWorker]:
+        with contextlib.closing(docker.from_env()) as client:
```
Same here, this should be triggered over an API call instead.

khoaguin replied:
Thanks @shubham3121, done!

shubham3121 committed:
Merge branch 'dev' into feat/container-workload-classes-repr

shubham3121 committed:
update protocol version json

khoaguin committed:
[syft/worker_image] WorkerPool queries running_workers and healthy_workers by
making an API call to the server

khoaguin committed:
[syft/worker_image] - add the worker.status service to api.services
- get SyftWorker's healthcheck by making the worker.status API call to the server

khoaguin commented:
Updated after having default worker pool in dev

![image](https://github.com/OpenMined/PySyft/assets/88959106/42ab5e81-158d-46a6-a8b9-ec160c3afdb2)

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/d332b20e-a537-42e7-8e1d-9cc8215aba2c)

khoaguin commented:
![image](https://github.com/OpenMined/PySyft/assets/88959106/2f8bddcb-99eb-421e-9f75-aeb1f1a1598b)

khoaguin committed:
[syft/worker_image] improve repr for `WorkerPool`
- add config as searchable attribute for `SyftWorkerImage`
- remove config as unique attribute for `SyftWorkerImage`

khoaguin committed:
[syft/worker_image] - only calling docker client to delete images if not using memory workers
- make some small changes to the api notebook to fix bugs and show repr

khoaguin committed:
[syft/worker_image] modify the api notebook to reflect code changes

khoaguin commented:
When using `os.environ["ORCHESTRA_DEPLOYMENT_TYPE"] = "container_stack"`, we get

![image](https://github.com/OpenMined/PySyft/assets/88959106/d41b87d1-acf8-4f39-bac2-aa3295897119)
![image](https://github.com/OpenMined/PySyft/assets/88959106/2412412c-d6c8-4991-8128-c2cabcc950b3)
![image](https://github.com/OpenMined/PySyft/assets/88959106/8a5460c2-90de-45cf-af8e-bd1cbe6911d7)
![image](https://github.com/OpenMined/PySyft/assets/88959106/a4408a79-4ad9-4891-91fd-4c88f5e4bfac)
![image](https://github.com/OpenMined/PySyft/assets/88959106/4fcf8d07-136c-4a35-8fef-6a069067253b)

rasswanth-s committed:
Merge branch 'dev' into feat/container-workload-classes-repr

shubham3121 committed:
clean up health check
- add a utility method for worker status and healthcheck update
- referesh status on resolving worker in worker pool
- running and healthy workers derived from workers
- skip status checks for in mem workers

Co-authored-by: rasswanth <43314053+rasswanth-s@users.noreply.github.com>

shubham3121 committed:
fix assignment of random tcp port for workers

shubham3121 committed:
set image identifier in case of in mem workers

shubham3121 reviewed (approved):
Awesome work @khoaguin !!! 🎉🎊
